### ISSUE TITLE
Template expressions containing non-trivial sizeof types

### ISSUE THREAD

**Issue by wsfulton** [swig/swig#2919]
```swig
%inline %{
#include <type_traits>
template <typename T>
class InternalHelper {
public:
// Original source: https://github.com/protocolbuffers/protobuf/blob/v20.2/src/google/protobuf/arena.h#L449-L458
    template <typename U>
        static char DestructorSkippable(const typename U::DestructorSkippable_*);
    template <typename U>
        static double DestructorSkippable(...);
        static short DestructorSkippable(int);

    typedef std::integral_constant<bool,
                      sizeof(DestructorSkippable<T>(static_cast<const T*>(0))) ==
                      sizeof(char)
//                  || std::is_trivially_destructible<T>::value
                  >
        is_destructor_skippable;
};
%}

%inline %{
bool tester() {
  InternalHelper<double>::is_destructor_skippable x;
  return x;
}
%}
```
results in:
```
example.i:18: Error: Syntax error in input(3).
```
Line 18 is the line containing:
```swig
                      sizeof(DestructorSkippable<T>(static_cast<const T*>(0))) ==
```
Seems like any non-trivial type in the `sizeof` fails to parse, even replacing line 18 with:
```swig
                      sizeof(DestructorSkippable(0)) ==
```

**Comment by ojwb** [comment]
We could probably handle arbitrary types in `sizeof()` via parser error recovery since we can find the closing `)` by skipping balanced `(`...`)` and SWIG doesn't actually need to parse the contents as the contents don't affect the type of the result of `sizeof()` (which is always `size_t`, though SWIG actually pretends it's `unsigned long`).  I'll give that a go.

That would also handle `sizeof` on arbitrary expressions, provided the whole expression is enclosed in parentheses (`sizeof <expr>` is valid without parentheses, but `sizeof <type>` isn't).  That would actually cover most real-world cases as it's very rare to omit the parentheses IME.

**Comment by ojwb** [comment]
I've also pushed a fix for a similar issue in `alignof(T)` (previously SWIG didn't treat `alignof` as a keyword and just handled `alignof(T)` the same as `foo(T)` without having seen a declaration of `foo`): a5de1914026984075fd49f1eb4a351d5d8ebb4a4

I took another quick look at resolving the restrictions on `sizeof EXPRESSION` (without parentheses).  It seems like unary operators might be better handled in `exprsimple` rather than `exprcompound` but moving them breaks some currently working cases.  Maybe this can work but it seems to need more poking than I have time for right now, and as I noted above it seems rare to omit the parentheses in practice.
